events.go

```diff
@@ -0,0 +1,182 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strconv"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/google/uuid"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/rs/zerolog"
+	tmrpcclient "github.com/tendermint/tendermint/rpc/client"
+	tmrpceventstream "github.com/tendermint/tendermint/rpc/client/eventstream"
+	rpchttp "github.com/tendermint/tendermint/rpc/client/http"
+	"github.com/tendermint/tendermint/rpc/coretypes"
+	tmjsonclient "github.com/tendermint/tendermint/rpc/jsonrpc/client"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+var (
+	queryEventBankTransfer = fmt.Sprintf(
+		"%s='%s' AND %s='%s'",
+		tmtypes.EventTypeKey,
+		tmtypes.EventTxValue,
+		"message.action",
+		"/cosmos.bank.v1beta1.MsgSend",
+	)
+)
+
+type EventCollector struct {
+	rpcClient tmrpcclient.Client
+	logger    zerolog.Logger
+	counter   *prometheus.CounterVec
+	up        *prometheus.GaugeVec
+}
+
+func NewEventCollector(tmRPC string, logger zerolog.Logger) (*EventCollector, error) {
+	httpClient, err := tmjsonclient.DefaultHTTPClient(tmRPC)
+	if err != nil {
+		return nil, err
+	}
+	// no timeout because we will continue fetching events continuously
+	rpcClient, err := rpchttp.NewWithClient(tmRPC, httpClient)
+	if err != nil {
+		return nil, err
+	}
+	transfersValueCounter := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name:        "cosmos_bank_transfer_amount",
+			Help:        "Number of tokens transferred in a transfer message",
+			ConstLabels: ConstLabels,
+		},
+		[]string{"denom", "sender", "recipient"},
+	)
+	upGauge := prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name:        "cosmos_bank_transfer_up",
+			Help:        "Whether the bank transfer event collector is up",
+			ConstLabels: ConstLabels,
+		},
+		[]string{"denom", "sender", "recipient"},
+	)
```

Comment:
Is `EventCollector.up` used anywhere? Also I think if we register the scrape target with prometheus, it should have this metric by default `up{}`

Reply:
no, i was thinking of adding one specifically for eventstream but because we panic fo rnonreco verable shutdown of streaming we'll  know via the `up` metric. I'll remove this artifact.